server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -442,6 +478,19 @@ private ThreadContextStruct putResponseHeaders(Map<String, List<String>> headers
 
         private ThreadContextStruct putResponse(final String key, final String value, final Function<String, String> uniqueValue) {
             assert value != null;
+            long curWrnHeaderSize = 0;
+            //check if we can add another warning header (max count or size within limits)
+            if (key.equals("Warning")) {
+                if (isWrnLmtReached) return this; //can't add warning headers - limit reached
+                if (maxWrnHeaderCount != -1) { //if count is NOT unbounded, check its limits
+                    int wrnHeaderCount = this.responseHeaders.containsKey("Warning") ? this.responseHeaders.get("Warning").size() : 0;
+                    if (wrnHeaderCount >= maxWrnHeaderCount) return addWrnLmtReachedHeader();
```

Comment:
Take care, the warning headers are de-duplicated below. The accounting should happen after de-duplication, otherwise we truncate when a warning header would not have been added because it is a duplicate of an existing warning header.